File: backend/schemas/suggestion.py
"""
Suggestion schemas for API request/response validation
"""
from pydantic import BaseModel, Field, ConfigDict
from datetime import datetime
from typing import Optional, List
from schemas.detection import DetectionResponse


class SuggestionBase(BaseModel):
    """Base suggestion schema"""
    detection_id: int
    suggested_label: int
    reason: str
    confidence: float = Field(ge=0, le=1)
    
    model_config = ConfigDict(protected_namespaces=())


class SuggestionCreate(SuggestionBase):
    """Schema for creating a suggestion"""
    pass


class SuggestionResponse(BaseModel):
    """Schema for suggestion response"""
    id: int
    detection_id: int
    suggested_label: int
    reason: str
    confidence: float
    status: str
    created_at: datetime
    reviewed_at: Optional[datetime] = None
    reviewer_notes: Optional[str] = None
    
    model_config = ConfigDict(from_attributes=True, protected_namespaces=())


class SuggestionWithDetection(SuggestionResponse):
    """Suggestion with full detection details"""
    detection_info: Optional[DetectionResponse] = None
    
    model_config = ConfigDict(from_attributes=True, protected_namespaces=())


class SuggestionGenerateRequest(BaseModel):
    """Request to generate suggestions"""
    dataset_id: int
    iteration: int = 1
    top_n: Optional[int] = Field(None, description="Limit to top N suggestions by priority")
    
    model_config = ConfigDict(protected_namespaces=())


class SuggestionGenerateResponse(BaseModel):
    """Response from generating suggestions"""
    dataset_id: int
    iteration: int
    suggestions_created: int
    total_detections: int
    message: Optional[str] = None
    
    model_config = ConfigDict(protected_namespaces=())


class SuggestionListResponse(BaseModel):
    """Response for listing suggestions"""
    suggestions: List[SuggestionResponse]
    total: int
    page: int
    page_size: int
    total_pages: int
    
    model_config = ConfigDict(protected_namespaces=())


class SuggestionUpdateRequest(BaseModel):
    """Request to update suggestion status"""
    status: str = Field(..., pattern="^(accepted|rejected|modified)$")
    reviewer_notes: Optional[str] = None
    custom_label: Optional[int] = None  
    
    model_config = ConfigDict(protected_namespaces=())


class SuggestionStatsResponse(BaseModel):
    """Statistics about suggestions"""
    dataset_id: int
    total_suggestions: int
    pending: int
    accepted: int
    rejected: int
    modified: int
    acceptance_rate: float
    
    model_config = ConfigDict(protected_namespaces=())

File: frontend/src/types/suggestion.ts
/**
 * Suggestion-related TypeScript types
 */

export interface Suggestion {
  id: number;
  detection_id: number;
  suggested_label: number;
  reason: string;
  confidence: number;
  status: SuggestionStatus;
  created_at: string;
  reviewed_at?: string;
  reviewer_notes?: string;
}

export interface SuggestionWithDetection extends Suggestion {
  detection_info?: {
    confidence_score: number;
    anomaly_score: number;
    priority_score: number;
    iteration: number;
  };
  sample_features?: number[];
  current_label?: number;
  predicted_label?: number;
  original_label?: number;
}

export interface SuggestionGenerateRequest {
  dataset_id: number;
  iteration?: number;
  top_n?: number;
}

export interface SuggestionGenerateResponse {
  dataset_id: number;
  iteration: number;
  suggestions_created: number;
  total_detections: number;
  message?: string;
}

export interface SuggestionListResponse {
  suggestions: Suggestion[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface SuggestionUpdateRequest {
  status: SuggestionReviewStatus; // Only reviewable statuses
  reviewer_notes?: string;
  custom_label?: number; // For 'modified' action
}

export interface SuggestionStats {
  dataset_id: number;
  total_suggestions: number;
  pending: number;
  accepted: number;
  rejected: number;
  modified: number;
  acceptance_rate: number;
}

export type SuggestionStatus = 'pending' | 'accepted' | 'rejected' | 'modified';

// Only statuses that can be set during review (excludes 'pending')
export type SuggestionReviewStatus = 'accepted' | 'rejected' | 'modified';

export interface SuggestionFilters {
  dataset_id?: number;
  iteration?: number;
  status?: SuggestionStatus;
  min_confidence?: number;
  page?: number;
  page_size?: number;
}